net/tstun/mtu.go

```diff
@@ -1,33 +1,171 @@
+// the MTU() method of tailscale's TUN wrapper.
+//
+// Maximum probed MTU: This is the largest MTU size that we send probe packets
+// for.
+//
+// Safe MTU: If the tailscale TUN MTU is set to this value, almost all packets
+// will get to their destination. Tailscale defaults to this MTU in the absence
+// of path MTU probe information or user MTU configuration. We may occasionally
+// find a path that needs a smaller MTU but it is very rare.
+//
+// Peer MTU: This is the path MTU to a peer's current best endpoint. It defaults
+// to the Safe MTU unless we have path MTU probe results that tell us otherwise.
+//
+// Initial MTU: This is the MTU tailscaled creates the TUN with. In order of
+// priority, it is:
+//
+// 1. If set, the value of TS_DEBUG_MTU clamped to a maximum of 65536
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 3. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+//
+// Current MTU: This the MTU of the tailscale TUN at any given moment
+// after TUN creation. In order of priority, it is:
+//
+// 1. The MTU set by the user via the OS, if it has ever been set
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 4. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+
+// TunMTU is used to describe MTUs for the tailscale TUN.
+type TunMTU uint32
+
+// WireMTU is used to describe MTUs for the underlying network devices.
+type WireMTU uint32
 
 const (
-	maxMTU     uint32 = 65536
-	defaultMTU uint32 = 1280
+	// Largest MTU we will consider for the Tailscale TUN. This is inherited
+	// from wireguard-go and can be surprisingly small; on Windows it is
+	// currently 2048 - 32 bytes.
+	maxTunMTU TunMTU = TunMTU(MaxPacketSize)
+	// A default "safe" MTU for the Tailscale TUN that we can use
+	// in the absence of other information such as path MTU probes.
+	safeTunMTU TunMTU = 1280
 )
 
-// DefaultMTU returns either the constant default MTU of 1280, or the value set
-// in TS_DEBUG_MTU clamped to a maximum of 65536.
-func DefaultMTU() uint32 {
-	// DefaultMTU is the Tailscale default MTU for now.
-	//
-	// wireguard-go defaults to 1420 bytes, which only works if the
-	// "outer" MTU is 1500 bytes. This breaks on DSL connections
-	// (typically 1492 MTU) and on GCE (1460 MTU?!).
-	//
-	// 1280 is the smallest MTU allowed for IPv6, which is a sensible
-	// "probably works everywhere" setting until we develop proper PMTU
-	// discovery.
-	tunMTU := defaultMTU
-	if mtu, ok := envknob.LookupUintSized("TS_DEBUG_MTU", 10, 32); ok {
-		mtu := uint32(mtu)
-		if mtu > maxMTU {
-			mtu = maxMTU
+// WireMTUsToProbe are the MTUs we are most likely to see in real world
+// networks. These are used to send discovery pings in packets padded
+// to exactly match the goal MTU.
+//
+// XXX These need adjusting before merging to main. What do you think?
+var WireMTUsToProbe = [...]WireMTU{
+	576,  // Smallest MTU for IPv4, probably useless?
```

Comment:
I don't think we should probe down to 576, we definitely don't support it today, and we would definitely need 576 + tunnel overhead to fit valid IPv4, which is our practical lower bound. I don't want to go down to 576+overhead either, though.

I want to suggest we don't probe below 1280, however, I also want to suggest:

- We have had a handful of WSL users trying to use Tailscale on Windows and WSL at the same time, then turn on an exit node on the Windows side and break themselves, as Tailscale can't carry Tailscale traffic today.
- Similar story, but for less easy to describe VM configurations.

So I'd like to suggest we go down to a minimum of 1280-tunnel overhead. This would allow us to fit exactly one layer of Tailscale over Tailscale. If someone sets up more layers in nested virt, I'm 100% ready to let that fail.

I would also be entirely comfortable with saying we only go as low as we do today, 1280.